src/HelixJobCreator.cs

```diff
@@ -126,6 +135,16 @@ public async Task<AgentInfoItem> CreateJob()
 
                 return new AgentInfoItem() { accepted = false };
             }
+            catch (OperationCanceledException ranOutOfTime) when (ranOutOfTime.CancellationToken == cancellationToken)
+            {
+                _logger.LogError($"Unable to complete request to create Helix job in specified timeout, attempting to cancel it.");
+                if (job != null && !string.IsNullOrEmpty(job.CorrelationId))
```

Comment:
How would we have gotten this value but also cancelled? Sending the job is the last thing this piece of code did.

Reply:
A `ThrowIfCancellationRequested` after the `SendAsync` call would hit this check. This check is necessary for the (rare) race condition where we send the request to helix right before azure devops cancels this call.

---
In reply to: [347669776](https://github.com/dotnet/arcade-pool-provider/pull/182#discussion_r347669776) [](ancestors = 347669776)

Reply:
Yeah, but... who cares? That's so crazy rare that it triggers in that single line of code. We should just remove that throw and return to the caller. It's really weird to cancel an Async call AFTER all the work has already been done... just return it.

Reply:
Especially since this is helix, there is a non-zero chance we've also already started working on the work item, so cancelling it won't do anything.  It just seems like a finicky piece of code that's so hard to hit that it will trigger like <1% of the time, meaning it's highly likely it will start misbehaving because of the rarity, and we'll just end up with a hard to diagnose bug that only triggers when this call takes EXACTLY 90 seconds to complete (not 89, because we'd have cancelled the send without getting the correlation ID, and not 91, because we've already returned the answer)

Reply:
If the cancellation token ever signals, AzDo is already in a bad state.  There's a small chance for a benefit and no chance for a negative side effect if we do cancel the job.  In the case of any queue that is currently backed up or scaled down, the chance of successful cancel goes up to about 100%; this in practice does happen quite a bit.

Reply:
The is risk. Code that runs rarely often has bugs. In increases the complexity of this class as well, making it more likely that someone will introduce an error in the future.  That cancellation will acquire small bugs because no one ever uses it, and when it does trigger 3 years from now, it will do something catastropihc.

Every line of code has a cost, and should pay for itself. Something that happens so rarely as to be effectively be unreachable costs more than it benefits.